### PR TITLE
feat(provider/tencent): Add tencentcloud support

### PR DESCRIPTION
**Pull request by zqfan**
As one of the major public cloud vendors in China, Tencent Cloud has been supporting spinnaker for a long while, now we decide to contribute to the spinnaker community, so here is this PR. But we don't know if it meets the community's requirement to merge, comments are welcomed, thanks!

This PR implements basic Bake process (we already support packer and released).

### CODE REVIEW

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy [hunk 1]
```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2015 Google, Inc.
```
Comment (jeyrschabu) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: Needs to be updated.

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import com.netflix.spinnaker.rosco.providers.CloudProviderBakeHandler
+import com.netflix.spinnaker.rosco.providers.tencent.config.RoscoTencentConfiguration
+import com.netflix.spinnaker.rosco.providers.util.ImageNameFactory
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+@Slf4j
+@Component
+public class TencentBakeHandler extends CloudProviderBakeHandler {
```
Comment (jeyrschabu) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: We are slowly deprecating the use of groovy in spinnaker services. We recommend either using `Java` or `Kotlin`.

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  BakeOptions getBakeOptions() {
+    new BakeOptions(
+      cloudProvider: BakeRequest.CloudProviderType.tencent,
+      baseImages: tencentBakeryDefaults?.baseImages?.collect { it.baseImage }
+    )
+  }
+
+  @Override
+  String produceProviderSpecificBakeKeyComponent(String region, BakeRequest bakeRequest) {
+//    if (!bakeRequest.vm_type) {
```
Comment (jeyrschabu) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: Please remove commented out code. A comment about why this should return null would be good though.

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+//
+//      return "$region:$vm_type:$enhancedNetworkingSegment"
+//    }
+    return null
+  }
+
+  @Override
+  def findVirtualizationSettings(String region, BakeRequest bakeRequest) {
+//    BakeRequest.VmType vmType = bakeRequest.vm_type ?: tencentBakeryDefaults.defaultVirtualizationType
+//
```
Comment (jeyrschabu) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: See comment on commented out code

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+      it.baseImage.id == bakeRequest.base_os
+    }
+
+    if (!operatingSystemVirtualizationSettings) {
+      throw new IllegalArgumentException("No virtualization settings found for '$bakeRequest.base_os'.")
+    }
+
+    def virtualizationSettings = operatingSystemVirtualizationSettings?.virtualizationSettings.find {
+      it.region == region //&& it.virtualizationType == vmType
+    }
```
Comment (jeyrschabu) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: same here

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy (hunk 2 above)
Comment (jeyrschabu) [reply rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: Unless the current structure absolutely forces you to use groovy

File: rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    logsContent.eachLine { String line ->
+      if (line =~ IMAGE_NAME_TOKEN) {
+        imageName = line.split(IMAGE_NAME_TOKEN).last()
+      } else if (line =~ "$region:") {
+        amiId = line.split("tencentcloud-cvm: Tencentcloud images\\($region: ").last().split("\\)").first()
+      }
+    }
+
+    return new Bake(id: bakeId, ami: amiId, image_name: imageName)
+  }
```
Comment (jeyrschabu) [inline rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: Is this applicable to your cloud provider?

Comment (jeyrschabu) [reply rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/tencent/TencentBakeHandler.groovy]: I was thrown off by the comment. This is probably fine

**Comment by zqfan** [comment]
@jeyrschabu Thanks for your careful review. Honestly speaking, we didn't know that the community prefer java more. Most of our code is groovy since we're learning example from existent code, For rosco it might be easy to change, for clouddriver and others with plenty LOC, might not.
This patch needs to catch up with latest code, and fix things according to your feedback., but I think this PR will be freezed for some time, my colleague will take over it.

**Comment by th3morg** [comment]
This PR has been waiting for updates for quite some time. We may consider closing it if there is not an update or someone to take over the necessary changes. @zqfan Will someone have time to work on the changes?

**Comment by zqfan** [comment]
I've already hand over all spinnake related work to another team. this pr can be closed now.

**Pull request closed without merging**